jrt/src/com/yahoo/jrt/TlsCryptoSocket.java

```diff
@@ -0,0 +1,253 @@
+        return channel;
+    }
+
+    @Override
+    public HandshakeResult handshake() throws IOException {
+        HandshakeState newHandshakeState = processHandshakeState(this.handshakeState);
+        log.fine(() -> String.format("Handshake state '%s -> %s'", this.handshakeState, newHandshakeState));
+        this.handshakeState = newHandshakeState;
+        return toHandshakeResult(newHandshakeState);
+    }
+
+    private HandshakeState processHandshakeState(HandshakeState state) throws IOException {
+        switch (state) {
+            case NOT_STARTED:
+                sslEngine.beginHandshake();
+                break;
+            case NEED_WRITE:
+                channelWrite();
+                break;
+            case NEED_READ:
+                channelRead();
+                break;
+            case COMPLETED:
+                return HandshakeState.COMPLETED;
+            default:
+                throw unhandledStateException(state);
+        }
+
+        while (true) {
+            switch (sslEngine.getHandshakeStatus()) {
+                case NOT_HANDSHAKING:
+                    if (wrapBuffer.bytes() > 0) return HandshakeState.NEED_WRITE;
+                    sslEngine.setEnableSessionCreation(false); // disable renegotiation
+                    handshakeDummyBuffer = null;
+                    SSLSession session = sslEngine.getSession();
+                    sessionApplicationBufferSize = session.getApplicationBufferSize();
+                    sessionPacketBufferSize = session.getPacketBufferSize();
+                    return HandshakeState.COMPLETED;
+                case NEED_TASK:
+                    sslEngine.getDelegatedTask().run();
+                    break;
+                case NEED_UNWRAP:
+                    if (wrapBuffer.bytes() > 0) return HandshakeState.NEED_WRITE;
+                    if (!handshakeUnwrap()) return HandshakeState.NEED_READ;
+                    break;
+                case NEED_WRAP:
+                    if (!handshakeWrap()) return HandshakeState.NEED_WRITE;
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected handshake status: " + sslEngine.getHandshakeStatus());
+            }
+        }
+    }
+
+    private static HandshakeResult toHandshakeResult(HandshakeState state) {
+        switch (state) {
+            case NEED_READ:
+                return HandshakeResult.NEED_READ;
+            case NEED_WRITE:
+                return HandshakeResult.NEED_WRITE;
+            case COMPLETED:
+                return HandshakeResult.DONE;
+            default:
+                throw unhandledStateException(state);
+        }
+    }
+
+    @Override
+    public int getMinimumReadBufferSize() {
+        return sessionApplicationBufferSize;
+    }
+
+    @Override
+    public int read(ByteBuffer dst) throws IOException {
+        verifyHandshakeCompleted();
+        int bytesUnwrapped = applicationDataUnwrap(dst);
```

Comment:
consider calling drain here to avoid returning a single frame even if multiple frames could be decoded